Laravel 7 native casts

atymic opened the pull request:
- [x] Added or updated tests
- [x] Added or updated the [README.md](../README.md)
- [x] Detailed changes in the [CHANGELOG.md](../CHANGELOG.md) unreleased section

WIP PR for implementing native eloquent casting.

Current issues:
- ~Currently, we let laravel do native casting, then we apply the enum cast after.~
~Ie. `'6'` cast by laravel -> `6` cast by us -> `EnumInstance`~

~With the new native casts, this isn't possible. I'm thinking we'll need to pass an additional attribute to the constructor to apply a native cast if required.~
~Kind of a bit messy, but not sure if there's a better way?~
Fixed in the latest commit :)


Also, it seems that natively cast attributes are not used in the `toArray()` which is a breaking change, so we need to see if we can return the object.

Edit: also need some way of only running this test on L7 in CI. Should be easy to do via ENV :)

BenSampo commented:
Worth taking a look at this PR for this feature:
https://github.com/laravel/framework/pull/32129

atymic commented:
I'll revisit this shortly :)

BenSampo commented:
Another relevant PR to to the framework:
https://github.com/laravel/framework/pull/32225

cerbero90 commented:
Thanks for this proposal @atymic, this feature would be great to have!

Do you think it is possible to:
- let `Enum` implement `Illuminate\Contracts\Database\Eloquent\Castable`
- return `EnumCast` in `castUsing()`

Introducing this feature may imply that a major version bump is needed for this package as custom casts are only supported in Laravel 7.

Previous Laravel applications working with versions `^1.0` of this package would fail as they wouldn't recognise the new interface.

Dependencies in `composer.json` and builds in `.travis.yml` should be updated accordingly too, or am I wrong @BenSampo?

atymic commented:
@cerbero90 

Yep, I agree it's worth a version bump for this feature.
@BenSampo i've updated the PR to use the `Castable` interface.

I've also added a way to do native casting (by adding a protected static property on the enum, with the native cast type)

Enum ends up looking like so:

```
final class UserType extends Enum
{
    protected static $nativeType = 'integer';

    const Administrator = 0;
    const Moderator = 1;
    const Subscriber = 2;
    const SuperAdministrator = 3;
}
```

and therefore, model can just be:

```
class NativeCastModel extends Model
{
    protected $casts = [
        'user_type' => UserType::class,
    ];

    protected $fillable = [
        'user_type',
    ];
}
```

cerbero90 commented:
Thanks for your work @atymic and for raising this issue in all my favourite Laravel packages (https://github.com/cknow/laravel-money/issues/48) 😄 

These are just my two cents and feel free to correct me, but would native casting be avoidable if the `$strict` flag in `hasValue()` defaulted to `false`?

This might be part of the breaking changes in the major version bump if @BenSampo thinks it's worth it.

It should be a curious edge case to have enums with
```php
const FOO = 1;
const BAR = '1';
```
and that case may be solved by calling `hasValue(1, true)`

atymic commented:
> These are just my two cents and feel free to correct me, but would native casting be avoidable if the $strict flag in hasValue() defaulted to false?

I hate the PHP type coercion, I think it's much better to be explicit.

atymic commented:
cc @spawnia who always has good thoughts :)

spawnia commented on src/Enum.php:
```diff
@@ -411,4 +413,15 @@ public static function getLocalizationKey(): string
+    {
+        $nativeType = null;
+
+        if (property_exists(get_called_class(), 'nativeType')) {
```
What is going on here? A comment might help.

atymic replied:
Replaced by new logic.

spawnia commented:
How about we do native casting by a simple function on the enum class?

```php
/**
 * Cast values before constructing an enum from them.
 *
 * You may need to overwrite this when using string values that are returned
 * from a raw database query or a similar data source.
 *
 * @param  mixed  $value  A raw value that may have any native type
 * @return mixed  The value cast into the type this enum expects
 */
public static function castNative($value)
{
    return $value;
}
```

Less overhead, less string-based magic.

spawnia commented:
> I don't think you should ever have an enum with 1 in both int and string forms. That's something I don't think is worth protecting from.

I agree, Enum's should be monomorphic. We should mention that explicitly as part of the docs.

cerbero90 commented:
> I don't think you should ever have an enum with 1 in both int and string forms. That's something I don't think is worth protecting from.

Agreed, what I meant is that my previous suggestion to default the flag to `false` would be problematic in presence of the edge case of values `1` and `'1'`, which is not recommended.

Would be great if we didn't have to specify a native cast but that would only be possible with type coercion.

cerbero90 commented:
What do you guys think about introducing the following method in `EnumCast`:

```php
/**
 * Retrieve the value that can be casted into Enum
 *
 * @param mixed $value
 * @return mixed
 */
protected function getCastableValue($value)
{
    // if the value exists in the enum, return it
    if ($this->enumClass::hasValue($value)) {
        return $value;
    }

    // if the value doesn't exist in the enum, return it and let it fail later
    if (!$this->enumClass::hasValue($value, false)) {
        return $value;
    }

    // return the value in the enum with the correct type
    foreach ($this->enumClass::getValues() as $enumValue) {
        if ($value == $enumValue) {
            return $enumValue;
        }
    }
}
```
So that `castEnum()` can call it to have the correct value to cast into Enum:
```php
protected function castEnum($value): ?Enum
{
    $enum = $this->enumClass;

    if ($value === null || $value instanceof Enum) {
        return $value;
    }

    $value = $this->getCastableValue($value);

    return $enum::getInstance($value);
}
```
This would let us:
- avoid defining a native type
- prevent performing native cast
- have an easier way to cast enums by providing the enum class name only

spawnia commented:
@cerbero90 so basically drop explicit casts and replace it with implicit PHP style type coercion?

cerbero90 commented:
@spawnia correct, the underlying logic remains the same but we implicitly provide the right type without asking the user

spawnia commented:
That is not guaranteed to produce a correct result. Let's suppose the Enum class has two possible values, which are returned as:

```
[true, false]
```

In the database the colums is defined as `ENUM ('true','false')`.

PHP's magic coercion would make it so that every value returned from the database would result in an enum with the value `true` being instantiated.

```
>>> true == 'true'
=> true                                                                                                                                                                                                                                       
>>> true == 'false'
=> true

cerbero90 commented:
Good point @spawnia. I can be wrong but to me that sounds like an edge case, nonetheless I agree that we should be able to handle it regardless.

What about using both solutions:
```php
protected function getCastableValue($value)
{
    // optionally cast the value into the expected type
    $value = $this->enumClass::castNative($value);

    // if the value exists in the enum, return it
    if ($this->enumClass::hasValue($value)) {
        return $value;
    }

    // if the value doesn't exist in the enum, return it and let it fail later
    if (!$this->enumClass::hasValue($value, false)) {
        return $value;
    }

    // return the value in the enum with the correct type
    foreach ($this->enumClass::getValues() as $enumValue) {
        if ($value == $enumValue) {
            return $enumValue;
        }
    }
}
```
That would solve the use case you mentioned by overriding your `castNative()` method with something like:
```php
public static function castNative($value)
{
    return $value === 'true';
}
```

spawnia commented:
@cerbero90 i like it, seems like a good compromise between convenience and offering control to the user.

atymic commented:
Awesome, thanks for the input @spawnia and @cerbero90 

I've pushed up that strategy and all the tests pass 🎉

cerbero90 commented:
Good stuff guys 👍 

@atymic do you mind changing the required PHP version from `~7.2.5` to `^7.2.5`?
This would let us support PHP >= 7.2.5 and < 8.0.0 instead of PHP >= 7.2.5 and < 7.3.0

atymic commented:
> @atymic do you mind changing the required PHP version from ~7.2.5 to ^7.2.5?
This would let us support PHP >= 7.2.5 and < 8.0.0 instead of PHP >= 7.2.5 and < 7.3.0

Fixed :)

spawnia commented on src/Casts/EnumCast.php:
```diff
@@ -59,28 +55,38 @@ protected function castEnum($value): ?Enum
+        }
+
+        // Try and find a value that can be type coerced into one that exists in the enum, returning if no matches
+        if (!$this->enumClass::hasValue($value, false)) {
```
I think this check might be redundant, the `foreach` loop below should take care of that.

spawnia commented on src/Casts/EnumCast.php:
```diff
@@ -0,0 +1,94 @@
+class EnumCast implements CastsAttributes
+{
+    /**@var string */
+    private $enumClass;
```
```suggestion
    protected $enumClass;
```

I think libraries should be extendable.

spawnia commented on src/Casts/EnumCast.php:
```diff
@@ -0,0 +1,94 @@
+     */
+    public function set($model, string $key, $value, array $attributes)
+    {
+        if ($value === null) {
```
Redundant check, happens in `castEnum`.

spawnia commented on src/Casts/EnumCast.php:
```diff
@@ -0,0 +1,94 @@
+
+        $enum = $this->enumClass;
+
+        if (!$value instanceof $enum) {
```
Redundant check, happens in `castEnum`.

spawnia commented on src/Casts/EnumCast.php:
```diff
@@ -0,0 +1,94 @@
+
+        $value = $this->getCastableValue($value);
+
+        return $enum::getInstance($value);
```
```suggestion
        return $this->enumClass::getInstance($value);
```

spawnia commented on src/Enum.php:
```diff
@@ -411,4 +413,23 @@ public static function getLocalizationKey(): string
+     * @param  mixed  $value  A raw value that may have any native type
+     * @return mixed  The value cast into the type this enum expects
+     */
+    public static function castNative($value)
```
A test case for overriding this function would be great.

atymic replied:
I'm just trying to think of a case that would actually test this. Can you think of one?

spawnia replied:
Override it's implementation in an Enum with something that produces a different cast than a default coercion would, e.g. `return ! $value;`.

atymic replied:
The issue is that it only does it on incoming values from the DB, not on the ones going back. Maybe we should support both directions?
I.e if you wanted to move from ints to strings, you could write a cast to transform the old values?

spawnia replied:
Oh right, totally forgot about that. Maybe we can have two methods then?

In terms of naming, this functionality is really only used on values of the database. How about `parseDatabase` and `serializeDatabase`?

That would make it clear when exactly this function will be called, especially that it is not involved when simply constructing an Enum. I can see how `new Enum('foo')` could be understood as "casting from native" - which is not what we are doing here.

atymic replied:
Have split them into different methods
[https://github.com/BenSampo/laravel-enum/pull/131/commits/44b25d12f6f281c5bbd5b7ca2d31dd5856fdba6e ](https://github.com/BenSampo/laravel-enum/pull/131/commits/44b25d12f6f281c5bbd5b7ca2d31dd5856fdba6e)

spawnia commented on composer.json:
```diff
@@ -19,17 +19,17 @@
+        "php": "^7.2.5",
         "hanneskod/classtools": "~1.0",
-        "illuminate/support": "5.4.*|5.5.*|5.6.*|5.7.*|5.8.*|^6.0|^7.0",
+        "illuminate/support": "^7.5",
```
```suggestion
        "illuminate/contracts": "^7.5",
        "illuminate/support": "^7.5",
```

There might be other Illuminate packages that are depended upon. Let's make them explicit.

spawnia commented:
Docs, Changelog and Upgrade guide are important for this PR. The upgrade guide should give concrete example of how to convert the casts, e.g.:

```diff
class MyModel extends Model
{
-   protected $enumCasts = [
+   protected $casts = [
        'foo' => Foo::class,
    ];
```

atymic commented:
I guess if we're doing a major version bump, maybe we should drop the existing casting method completely since it's replaced by custom casts?

Or we can leave it (and deprecate it)

spawnia commented:
@atymic i think we should try and have a release that contains both in order to allow incremental adoption. If we can introduce that in the current version, that would be great.

I think we should remove `$enumCasts` after a grace period that allows for possible bugs or edge cases in the new native casts to be fixed.

spawnia commented on README.md:
````diff
@@ -449,6 +438,38 @@ $example->user_type = UserType::Moderator;
+}
+```
+
+### Upgrading from Laravel Enum 1.x
````
I recommend we put this into a dedicated `UPGRADE.md` and maybe link there from the README. That makes it easy to find when looking at the repo.

spawnia commented on README.md:
````diff
@@ -449,6 +438,38 @@ $example->user_type = UserType::Moderator;
+move the casts from `$enumCasts` to `$casts`. 
+
+```diff
+class MyModel extends Model
````
```suggestion
-use BenSampo\Enum\Traits\CastsEnums;

class MyModel extends Model
```

netpok commented:
I would say if it's not hurting anything else then it should be only dropped when the LTS is no longer supported or there is an other change that actively requires the version bump.

cerbero90 commented:
Hey guys, any updates on this feature? Let me know if you need help

shaffe-fr commented:
Hey, I can help if you need help. 
I've been using this PR on my project for a few weeks and I encounter no problems.

atymic commented:
I'll clean this up and get it ready to merge shortly :)

If you'd like to help, could you write the `UPGRADE.md` doc?

Thanks @shaffe-fr

atymic commented:
Ok, have added docs and made the required changes. Thanks @shaffe-fr but no help needed :)

atymic commented:
@BenSampo should be good to go for review now :)

spawnia commented on README.md:
```diff
@@ -412,17 +409,9 @@ class Example extends Model
 {
     use CastsEnums;
```
Agreed, this should definitely be removed.

spawnia commented on README.md:
````diff
@@ -449,6 +439,23 @@ $example->user_type = UserType::Moderator;
 ```
 
+### Casting underlying native types
+Many databases return everything as strings (for example, an integer may be returned as the string `'1'`). To reduce fricton for users of the library we'll use type coercion to
````
```suggestion
Many databases return everything as strings (for example, an integer may be returned as the string `'1'`).
To reduce friction for users of the library, we use type coercion to
```

spawnia commented on README.md:
```diff
@@ -449,6 +439,23 @@ $example->user_type = UserType::Moderator;
+    const Administrator = 0;
+    const Moderator = 1;
+    
+    public static function castNative($value)
```
This example is outdated and should use `(parse|serialize)Database`.

spawnia commented on src/Enum.php:
```diff
@@ -411,4 +413,37 @@ public static function getLocalizationKey(): string
+        return $value;
+    }
+
+    public static function castUsing()
```
```suggestion
    public static function castUsing(): CastsAttributes
```

spawnia commented on UPGRADE.md:
````diff
@@ -0,0 +1,23 @@
+Trait based casting is still present, but is now deprecated and will be removed in the next major version.
+
+```diff
+class MyModel extends Model
+{
````
```suggestion
--use BenSampo\Enum\Traits\CastsEnums;

class MyModel extends Model
{
-    use CastsEnums;

```

shaffe-fr commented:
I've been using this with a custom composer repo, this is very smooth and clean. Only problem I got was caused by a problem in laravel that will be fixed next release.
Great contribution 😊

atymic commented:
@BenSampo sorry to bother you mate, let me know if there's any changes you'd like :)

BenSampo commented:
All looks good. Awesome job on this @atymic and well supported by @spawnia and @cerbero90 - cheers guys.
